### ISSUE TITLE
Feature request: Filter argument to `walkdir`

### ISSUE THREAD

**Issue by jakobnissen** [JuliaLang/julia#46312]
When traversing a large directory with sub-directories, there is currently no way of preventing the `walkdir` iterator recursing into an unwanted directory.
This comes up often:
* Perhaps the directory is irrelevant, but 99% of the subdirectories reside there, so most the time is wasted traversing here
* Perhaps a particular directory is mounted to an unresponsive drive/network, and so would stall the iterator for minutes if encountered
* Perhaps the user does not have access to an irrelevant sub-directory and so it would throw an error, and the user does not want to change the `onerror` keyword, because the user actually _does_ want to throw an error if the desired sub-directory is not accessible.

We could change it with a `filter` keyword, or alternatively as the first positional argument to allow `do`-syntax. Any subdirectory `d` for which the filter function `f(d) === false` would be skipped and never descended into.

**Comment by mgkuhn** [comment]
I agree that a directory-pruning filter is an essential missing feature of [walkdir](https://docs.julialang.org/en/v1/base/file/#Base.Filesystem.walkdir). (speaking as a regular user of the equivalent Perl module [`File::Find`](https://perldoc.perl.org/File::Find), which has undergone a lot of performance tuning)

However, I wonder if the proposed interface won't hinder performance:

Since `walkdir` does already prepare an array of subdirectories for each directory that it visits, wouldn't it be more efficient if the proposed `filter` function is not called with `filter(path::String)` for each directory visited, but instead is called as `filter(parent_path::String, dirs::Vector{String})` and is then expected to return a filtered version of `dirs::Vector::{String}` that contains only the subdirectories to be recursed into. Often, filters will operate only on the directory name, rather than the entire path. (A typical example would be to suppress “invisible” `.*` files. Therefore, having the directory name already separated out can save a lot of CPU cycles otherwise needed on string operations.

In addition, there could also be another kwarg parameter `chdir::Boolean=false` that can be enabled such that when `filter` is called, the current working directory is already set to `parent_path`, such that if the filter wants to test some property of the file (e.g., access permissions is a common case), it can hand over to the OS just the short relative filename.

Such a chdir option would also be particularly useful in case we add a `do` version of `walkdir`, which instead of returning an iterator calls a function for each file visited, in which case the act of that user-supplied function doing something to the file can be more efficient if the current working directory is the parent directory. Some operating systems handle file-system requests a lot faster if you give them short relative paths relative to the current working directory, an that's why Perl's [`File::Find`](https://perldoc.perl.org/File::Find) chose that kind of interface.

**Comment by mgkuhn** [comment]
Another important filter criterion is whether a file has already been visited before. While `walkdir` has an option to allow following symbolic links, it doesn't state or provide any control over what happens in case such symbolic links cause the same file or directory to be visited multiple times.

**Comment by jakobnissen** [comment]
I don't think the efficiency matters here. The cost of the string operation related to `joinpath` is insignificant compared to the system calls of `walkdir`.
But I'm not well versed in systems programming, so I think these concerns are better answered by someone more experienced than I

**Comment by mgkuhn** [comment]
Don't forget the time the kernel needs to undo the `joinpath` at the other end. Linux tends to be pretty good at this (through aggressive caching of path-prefix lookups), Windows NTFS not so much, as many Windows `git` users will be well aware off.
